src/analysis/lattice.h

```diff
@@ -1,55 +1,108 @@
+template<typename Element>
 constexpr bool has_isBottom =
-  std::is_invocable_r<bool, decltype(T::isBottom), const T&>::value;
+  std::is_invocable_r<bool, decltype(&Element::isBottom), Element>::value;
+
+template<typename Lattice>
+constexpr bool is_lattice = has_getBottom<Lattice>&& has_compare<Lattice>&&
+  has_makeLeastUpperBound<typename Lattice::Element>&& has_isTop<
+    typename Lattice::Element>&& has_isBottom<typename Lattice::Element>;
+
+// Represents a powerset lattice which is constructed from a finite set which
+// can be represented by a bitvector. Set elements are represented by
+// FinitePowersetLattice::Element, which represents members present in each
+// element by bits in the bitvector.
+class FinitePowersetLattice {
+
+  // The size of the set that the powerset lattice was created from. This is
+  // equivalent to the size of the Top lattice element.
+  size_t setSize;
+
+public:
+  FinitePowersetLattice(size_t setSize) : setSize(setSize) {}
+
+  // This represents an element of a powerset lattice. The element is itself a
+  // set which has set members. The bitvector tracks which possible members of
+  // the element are actually present.
+  class Element {
+    // If bitvector[i] is true, then member i is present in the lattice element,
+    // otherwise it isn't.
+    std::vector<bool> bitvector;
 
-template<typename T>
-constexpr bool is_lattice =
-  has_compare<T>&& has_getLeastUpperBound<T>&& has_isTop<T>&& has_isBottom<T>;
+    // This constructs a bottom element, given the lattice set size. Used by the
+    // lattice's getBottom function.
+    Element(size_t latticeSetSize) : bitvector(latticeSetSize) {}
 
-// Represents a powerset lattice element (i.e. a set) as a bitvector. A true
-// means that an element is present in the set.
-template<size_t N> struct BitsetPowersetLattice {
-  std::bitset<N> value;
+  public:
+    Element(Element&& source) = default;
+    Element(Element& source) = default;
 
-  static BitsetPowersetLattice<N> getBottom();
-  static bool isTop(const BitsetPowersetLattice<N>& element);
-  static bool isBottom(const BitsetPowersetLattice<N>& element);
+    Element& operator=(Element&& source) = default;
+    Element& operator=(const Element& source) = default;
+
+    // Counts the number of members present the element itself. For instance, if
+    // we had {true, false, true}, the count would be 2. O(N) operation which
+    // iterates through the bitvector.
+    size_t count();
+
+    bool get(size_t index) { return bitvector[index]; }
+    void set(size_t index, bool value) { bitvector[index] = value; }
+
+    bool isTop() { return count() == bitvector.size(); }
+    bool isBottom() { return count() == 0; }
+
+    // Calculates the LUB of this element with some right element and sets
+    // this element to the LUB in place. Returns true if this element before
+    // this method call was different than the LUB.
+    bool makeLeastUpperBound(const Element& right);
```

Comment:
```suggestion
    bool makeLeastUpperBound(const Element& other);
```
We have the convention that when there are two items with names it is `left, right` as you did below. But when there is just another one in addition to `*this` then we use `other`.

Reply:
Fixed.